From eb36bf5adb44978d47f48087ada2704f7eb8bbf1 Mon Sep 17 00:00:00 2001
From: Larry Finger <Larry.Finger@lwfinger.net>
Date: Wed, 26 Feb 2014 22:47:48 -0600
Subject: [PATCH 210/210] Fix kernel oops if firmware is not available

This patch also copies the firmware when doing an install.

Signed-off-by: Larry Finger <Larry.Finger@lwfinger.net>
---
 Makefile                |  2 ++
 hal/rtl8192d_hal_init.c | 14 ++++++--------
 2 files changed, 8 insertions(+), 8 deletions(-)

--- a/drivers/staging/rtl8192du/Makefile
+++ b/drivers/staging/rtl8192du/Makefile
@@ -168,6 +168,8 @@ strip:
 
 install:
 	install -p -m 644 $(MODULE_NAME).ko  $(MODDESTDIR)
+	mkdir -p /lib/firmware/rtlwifi
+	cp -n rtl8192dufw*.bin /lib/firmware/rtlwifi/.
 	/sbin/depmod -a ${KVER}
 
 uninstall:
--- a/drivers/staging/rtl8192du/hal/rtl8192d_hal_init.c
+++ b/drivers/staging/rtl8192du/hal/rtl8192d_hal_init.c
@@ -276,12 +276,12 @@ static int _FWInit(struct rtw_adapter *a
 	return _FAIL;
 }
 
-static int get_fw_from_file(struct rtw_adapter *adapter)
+static bool get_fw_from_file(struct rtw_adapter *adapter)
 {
 	struct dvobj_priv *dvobj = adapter_to_dvobj(adapter);
 	struct device *device = dvobj_to_dev(dvobj);
 	const struct firmware *fw;
-	int rtstatus = _SUCCESS;
+	bool rtstatus = false;
 #ifdef CONFIG_WOWLAN
 	const char *fw_name = "rtlwifi/rtl8192dufw_wol.bin";
 #else
@@ -289,13 +289,12 @@ static int get_fw_from_file(struct rtw_a
 #endif /* CONFIG_WOWLAN */
 
 	if (request_firmware(&fw, fw_name, device))
-		return _FAIL;
+		return false;
 	if (!fw) {
 		pr_err("Firmware %s not available\n", fw_name);
-		return _FAIL;
+		return false;
 	}
 	if (fw->size > FW_8192D_SIZE) {
-		rtstatus = _FAIL;
 		pr_err("Firmware size exceeds 0x%x. Check it.\n",
 		       FW_8192D_SIZE);
 		goto exit;
@@ -303,20 +302,19 @@ static int get_fw_from_file(struct rtw_a
 
 	adapter->firmware = kmalloc(sizeof(struct rt_firmware_92d), GFP_KERNEL);
 	if (!adapter->firmware) {
-		rtstatus = _FAIL;
 		goto exit;
 	}
 	adapter->firmware->buffer = vzalloc(fw->size);
 	if (!adapter->firmware->buffer) {
 		kfree(adapter->firmware);
 		adapter->firmware = NULL;
-		rtstatus = _FAIL;
 		goto exit;
 	}
 	memcpy(adapter->firmware->buffer, fw->data, fw->size);
 	adapter->firmware->length = fw->size;
 	pr_info("r8192du: Loaded firmware file %s of %d bytes\n",
 		fw_name, adapter->firmware->length);
+	rtstatus = true;
 exit:
 	release_firmware(fw);
 	return rtstatus;
@@ -347,7 +345,7 @@ int FirmwareDownload92D(struct rtw_adapt
 		return _FAIL;
 
 	if (!adapter->firmware) {
-		if (get_fw_from_file(adapter)) {
+		if (!get_fw_from_file(adapter)) {
 			rtStatus = _FAIL;
 			adapter->firmware = NULL;
 			goto Exit;
